### PR TITLE
Update JIT sources to clang-format/clang-tidy 17.0.6

### PR DESCRIPTION
**Pull request by BruceForstall**
The version of clang-format/clang-tidy we use in the JIT is very old and needs to be updated: see https://github.com/dotnet/runtime/issues/59545. One reason is to allow support for jit-format on arm64 Mac, and possibly arm64 Linux. Another is simply to better support newer language features and fix bugs.

This PR shows an updated clang-format configuration file that gets us about as close as possible to the formatting we currently have (it is impossible to get it 100% the same). There are some nice improvements here.

### CODE REVIEW

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -1332,72 +1332,72 @@ AssertionIndex Compiler::optCreateAssertion(GenTree*         op1,
...
-                CNS_COMMON:
-                {
-                    //
-                    // Must either be an OAK_EQUAL or an OAK_NOT_EQUAL assertion
-                    //
-                    if ((assertionKind != OAK_EQUAL) && (assertionKind != OAK_NOT_EQUAL))
-                    {
-                        goto DONE_ASSERTION; // Don't make an assertion
-                    }
+CNS_COMMON:
```
Comment (BruceForstall) [inline src/coreclr/jit/assertionprop.cpp:1335]: One new things is labels are left-justified, and it appears that code blocks are then aligned to the preceding label. That's a bit weird.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib @jkoritzinsky PTAL

**Comment by BruceForstall** [comment]
Related: https://github.com/dotnet/runtime/issues/71983

cc @a74nh 

Looks like the 17.0.6 release has some aarch64 Linux binaries: https://github.com/llvm/llvm-project/releases/tag/llvmorg-17.0.6

**Comment by BruceForstall** [comment]
I'd rather go to the newest possible, and 17.0.6 was the latest that also had built Ubuntu binaries (as well as osx-arm64, windows-x64). I don't think we should tie the JIT use of clang-format/clang-tidy to any other use of LLVM; that just complicates dependencies and our ability to update.

**Comment by BruceForstall** [comment]
Note that this change requires https://github.com/dotnet/jitutils/pull/408.

All the pieces are in place where we could take this change now. I've tested jit-format on linux-x64 (Ubuntu 22.04.4 LTS), win-x64, and osx-arm64.

File: src/coreclr/jit/assertionprop.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/assertionprop.cpp:1335]: Looks like this setting: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentgotolabels

File: src/coreclr/jit/block.cpp [hunk 2]
```diff
@@ -231,9 +236,13 @@ FlowEdge* Compiler::BlockPredsWithEH(BasicBlock* blk)
...
-                    assert(filterBlk->VisitEHEnclosedHandlerSecondPassSuccs(this, [blk](BasicBlock* succ) {
-                        return succ == blk ? BasicBlockVisit::Abort : BasicBlockVisit::Continue;
-                    }) == BasicBlockVisit::Abort);
+                    assert(filterBlk->VisitEHEnclosedHandlerSecondPassSuccs(this,
+                                                                            [blk](BasicBlock* succ)
+                                                                            {
+                                                                                return succ == blk
+                                                                                           ? BasicBlockVisit::Abort
+                                                                                           : BasicBlockVisit::Continue;
+                                                                            }) == BasicBlockVisit::Abort);
```
Comment (SingleAccretion) [inline src/coreclr/jit/block.cpp:245]: This lambda formatting change looks like a significant regression... Maybe turn https://clang.llvm.org/docs/ClangFormatStyleOptions.html#bracewrapping, `BeforeLambdaBody`, to `false`? Or https://clang.llvm.org/docs/ClangFormatStyleOptions.html#lambdabodyindentation to `OuterScope`?

Comment (BruceForstall) [reply src/coreclr/jit/block.cpp]: I set both of these as you suggest, and it looks pretty good to me.

File: src/coreclr/jit/assertionprop.cpp (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/jit/assertionprop.cpp]: I set `IndentGotoLabels: true` and this seems to be fixed.

**Comment by BruceForstall** [comment]
I updated the formatted code if you want to look, or look again.

**Comment by BruceForstall** [comment]
@SingleAccretion Thanks for looking!

> 1) Code comments align together with those from #ifdef

That one's odd. I think we'll have to work around it with additional whitespace.

> 2) Code aligns together with an if that is meant to be conditionally else if:

I think the best way to fix this is to avoid having "else" and "if" in separate `#ifdef` blocks, which IMO is already a readability problem.

> 3) Long comments on macros induce multi-line macros:

It doesn't happen too often, and I think the fix is to put non-trivial (i.e., multi-line) comments before the macros, e.g.,
```
// Set this to '1' to enable warning C4820 "'bytes' bytes padding added after
// construct 'member_name'" on interesting structs/classes
#define CHECK_STRUCT_PADDING 0
```

Related: I'd like to reformat a lot of jitconfigvalues.h this way, as it has far too many weirdly-formatted and multi-line comments.

> There is also one miscellaneous change: closing braces after namespaces now get a comment:

Looks useful to me (I hadn't notice that one; we don't use namespaces too often).

**Comment by BruceForstall** [comment]
This is ready to go. It requires https://github.com/dotnet/jitutils/pull/408 as well.

I tested the CI format job with https://github.com/dotnet/runtime/pull/100565, which is the same as this PR but pulled the jitutils bootstrap.cmd/sh from my jitutils PR branch.

Note that osx-x64 is no longer supported for formatting (but osx-arm64 is!). (It might be possible if you build your own clang 17.0.6 osx-x64 clang-format/clang-tidy?)

I'd like someone to pull down this and the corresponding jitutils PR on an osx-arm64 machine to verify it works for them, as well.

@AndyAyersMS @jakobbotsch @dotnet/jit-contrib PTAL

**Review by AndyAyersMS — APPROVED** [review]
Thanks for doing this. The new format looks better.

**Comment by BruceForstall** [comment]
/ba-g Formatting job failures are expected due to required cross-repo coordinated checkin

**Pull request merged**